### ISSUE TITLE
[Blazor][Wasm] Using roles with ApiAuthorization + Blazor

### ISSUE THREAD

**Issue by javiercn** [dotnet/AspNetCore.Docs#17649]
[EDIT by guardrex to add the topic metadata]

Feel free to dupe this issue with some other issue, I couldn't find a specific one.

Content is here:
## Steps for using roles with ApiAuthorization + Client-side Blazor
* Configure Identity to use roles by calling AddRoles
  ```csharp
  services.AddDefaultIdentity<ApplicationUser>(options => options.SignIn.RequireConfirmedAccount = true)
      .AddRoles<IdentityRole>()
      .AddEntityFrameworkStores<ApplicationDbContext>();
  ```
* Configure identity server to put the role claim into the id token and the access token and prevent the default mapping for roles in the JwtSecurityTokenHandler.
  ```csharp
  services.AddIdentityServer()
      .AddApiAuthorization<ApplicationUser, ApplicationDbContext>(options => {
          options.IdentityResources["openid"].UserClaims.Add("role");
          options.ApiResources.Single().UserClaims.Add("role");
      });
  
  // Need to do this as it maps "role" to ClaimTypes.Role and causes issues
  JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Remove("role");
  ```
* On your blazor application use `[Authorize(Roles = "admin")]` or any other role your app defines.
* On your protected resource APIs use `[Authorize(Roles = "admin")]` or any other role your app defines.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 17515ff7-32a6-c6f7-0425-535775cd6e28
* Version Independent ID: 1416abd7-0b23-ca93-62e7-59d2532d4275
* Content: [Secure an ASP.NET Core Blazor WebAssembly hosted app with Identity Server](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-identity-server?view=aspnetcore-3.1#feedback)
* Content Source: [aspnetcore/security/blazor/webassembly/hosted-with-identity-server.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/blazor/webassembly/hosted-with-identity-server.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-security**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
I'm putting up an issue this afternoon (Tuesday) to cover Azure AD roles and groups in WASM topics that use Azure. The two issues can be worked together.

**Comment by guardrex** [comment]
Related to https://github.com/dotnet/AspNetCore.Docs/issues/17317. Work these together.

cc: @denmitchell ... this issue comment may help you until I work the issue. I won't close your issue or call this an exact duplicate because you also asked about *policies*. We might need a little something for that scenario (i.e., set up a policy based on claims provided by IS).

**Comment by caleblloyd** [comment]
I have been scouring documentation and source code for a few hours trying to get Roles and Policies to work in Blazor WebAssembly..

The documentation listed in this issue took care of the Server portion.  But I still had to add the following additional configuration to the Client project to get this working with the `<AuthorizeView Role="roleName">` component on the Blazor WebAssembly client.

1. Add Class to configure `options.UserOptions.RoleClaim` on the Client

```
using Microsoft.AspNetCore.Components.WebAssembly.Authentication;
using Microsoft.Extensions.Options;

namespace App.Client.Services
{
    public class ApiAuthorizationOptionsConfiguration
        : IPostConfigureOptions<RemoteAuthenticationOptions<ApiAuthorizationProviderOptions>>
    {
        public void Configure(RemoteAuthenticationOptions<ApiAuthorizationProviderOptions> options)
        {
            options.UserOptions.RoleClaim ??= "role";
        }

        public void PostConfigure(string name, RemoteAuthenticationOptions<ApiAuthorizationProviderOptions> options)
        {
            if (string.Equals(name, Options.DefaultName))
            {
                Configure(options);
            }
        }
    }
}
```

2. Modify the `Program.cs` file to call `ApiAuthorizationOptionsConfiguration` and configure the `role` claim.

```
using App.Client.Services;
...

namespace App.Client
{
    public class Program
    {
        public static async Task Main(string[] args)
        {
            ...

            builder.Services.TryAddEnumerable(
                ServiceDescriptor.Singleton<
                    IPostConfigureOptions<RemoteAuthenticationOptions<ApiAuthorizationProviderOptions>>,
                    ApiAuthorizationOptionsConfiguration>());

            ...
        }
    }
}
```

Interested to hear if there is an easier way to enable Roles on the client.  Or if this is the easiest way, could you add it to the documentation?  I've also put these findings into a  [StackOverflow](https://stackoverflow.com/a/61153225/1419658) post

**Comment by guardrex** [comment]
@caleblloyd ... In your testing, did you merely try (in the client app and with a `role` claim coming down) ...

```csharp
builder.Services.AddApiAuthorization();
builder.Services.AddApiAuthorization(options =>
{
    options.UserOptions.RoleClaim = "role";
});
```

**Comment by caleblloyd** [comment]
Yes, I did try that first.  It causes the call to:

```
http://localhost:5000/_configuration/App.Client
```

To change to this, which results in a 404

```
http://localhost:5000/_configuration/Microsoft.AspNetCore.Components.WebAssembly.Authentication
```

It appears that there was already an issue for this and it has been fixed in preview4: https://github.com/dotnet/aspnetcore/issues/19854

So if that fix is good, it's probably best to just use your snippet in the documentation, much simpler!

**Comment by caleblloyd** [comment]
I can confirm that your workaround with the double call to `AddApiAuthorization` does work for preview3:

```
builder.Services.AddApiAuthorization();
builder.Services.AddApiAuthorization(options =>
{
    options.UserOptions.RoleClaim = "role";
});
```

Thanks!

**Comment by guardrex** [comment]
I'm new to IS, so I'm **_all for it being simple_** to get the various scenarios covered.

I also noticed an SO post where a user used the Profile Service to do what @javiercn did in just a few lines of code. I wonder if IS users will find that useful for other reasons (i.e., they need Profile Service anyway for other purposes). We might end up showing both approaches. The Profile Service way is ...

```csharp
using IdentityModel;
using IdentityServer4.Models;
using IdentityServer4.Services;
using System.Threading.Tasks;

namespace HostedIdentityServer.Server
{
    public class MyProfileService : IProfileService
    {
        public MyProfileService()
        {
        }

        public Task GetProfileDataAsync(ProfileDataRequestContext context)
        {
            var roleClaims = context.Subject.FindAll(JwtClaimTypes.Role);
            context.IssuedClaims.AddRange(roleClaims);
            return Task.CompletedTask;
        }

        public Task IsActiveAsync(IsActiveContext context)
        {
            return Task.CompletedTask;
        }
    }
}
```

```csharp
// using IdentityServer4.Services;

// Use the normal IS services add (i.e., the OOB template way) ...
services.AddIdentityServer()
    .AddApiAuthorization<ApplicationUser, ApplicationDbContext>();

// Add the Profile Service ...
services.AddTransient<IProfileService, MyProfileService>();
```

... also gets the `role` claim down to the client-side app. Having said that tho ... I'm not sure about side-effects or **_gotchas_** with that approach based on my being new to the whole scenario.

**Comment by AlbertoPa** [comment]
The profile approach does implicitly what you achieved with the AddApiAuthorization command in the client, by adding the claim to the token sent back to the server with the request, so they both should be OK. 

If nothing is done on the client, with the approach provided at the top of this issue, the user claims sent from the server are:

![image](https://user-images.githubusercontent.com/423032/79057948-c85a7c80-7c2c-11ea-9438-1beaf8b12e50.png)

The reason why adding the role claim in the client is not needed if the profile approach is used is because the token contains it already (highlighted in yellow; ignore the other claims, I added them to do other tests, see here for the whole set of steps: https://github.com/dotnet/AspNetCore.Docs/issues/17317#issuecomment-609187899):

![image](https://user-images.githubusercontent.com/423032/79057956-d27c7b00-7c2c-11ea-8e22-16e1499ad7be.png)

So, when the client sends the request, just shooting back the token, all is in place. I don't know if this is the best approach, but so far, while waiting for official guidance to check against, it has worked OK. 

**Note**: in the post I linked, I did not mention that also with the profile approach, it is necessary to clear the default mapping. I did this with

`JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();`

in the server Startup.cs.

**Comment by guardrex** [comment]
> with the profile approach, it is necessary to clear the default mapping

I didn't do that in my test with the Profile Service. Did you mean it's necessary with the client-added `role` claim approach (as Javier points out in the opening remark)?

**Comment by AlbertoPa** [comment]
> I didn't do that in my test with the Profile Service. Did you mean it's necessary with the client-added `role` claim approach (as Javier points out in the opening remark)?

I think it is necessary for how I did it in the page I linked (all claims are being remapped in the implementation I posted). I have not tried only mapping the role with the profile.

**Comment by hakenr** [comment]
The workarounds described above does not help if there is more than one role provided in the JWT token. In such case the `role` section of the token looks like
```json
  "role": [
    "UserSettingsAdministrator",
    "SystemAdministrator"
  ]
```
...which is not correctly handled by Microsoft.AspNetCore.Components.WebAssembly.Authentication package. You can "hack" (post-fix) this issue by overriding and replacing the `AccountClaimsPrincipalFactory<TAccount>` service with this one:
```csharp
public class RolesAccountClaimsPrincipalFactory : AccountClaimsPrincipalFactory<RemoteUserAccount>
{
	public RolesAccountClaimsPrincipalFactory(IAccessTokenProviderAccessor accessor) : base(accessor)
	{
		// NOOP
	}

	public override ValueTask<ClaimsPrincipal> CreateUserAsync(RemoteUserAccount account, RemoteAuthenticationUserOptions options)
	{
		var roles = account.AdditionalProperties["role"] as JsonElement?;
		if (roles?.ValueKind == JsonValueKind.Array)
		{
			account.AdditionalProperties.Remove("role");
			var claimsPrincipal = base.CreateUserAsync(account, options).Result;

			foreach (JsonElement element in roles.Value.EnumerateArray())
			{
				((ClaimsIdentity)claimsPrincipal.Identity).AddClaim(new Claim("role", element.GetString()));
			}
			return new ValueTask<ClaimsPrincipal>(claimsPrincipal);
		}

		return base.CreateUserAsync(account, options);
	}
}
```
Register this new service right before calling the `AddApiAuthorization()`:
```csharp
			builder.Services.AddScoped(typeof(AccountClaimsPrincipalFactory<RemoteUserAccount>), typeof(RolesAccountClaimsPrincipalFactory));
builder.Services.AddApiAuthorization();
```
(As `account.AdditionalProperties` is a dictionary, you are not able to split the roles right there.)

**Comment by guardrex** [comment]
@hakenr ... I saw a similar change for the `roles` claim from AAD. It used to be (Preview 2) that the value of the `roles` claim was just a comma-separated list of roles in the claim value.

Anyway ... that's not directly related except to say that if the WASM auth gets a `roles` claim with an array like that that it can't parse it and thus will fail for `IsInRole`. It might disturb policy checks as well.

I'll take a further look at this in the next day or two (hopefully Monday) and report back.

**Comment by guardrex** [comment]
Thanks for that code @hakenr ... that was a nice basis for a user factory fix to some AAD security groups and user-defined roles code that I was working with at https://github.com/dotnet/AspNetCore.Docs/issues/17660#issuecomment-616710615. Eventually, I hope we can have happy path documentation for assigning AAD security groups/roles, then devs can easily use the app's auth approaches ...

* `<AuthorizeView Roles="customrole">`
* `@attribute [Authorize(Roles = "customrole")]`
* `if (user.IsInRole("customrole")) { ... }`

**Comment by jvallejoa** [comment]
Hi programmers, I also used to get custom Claims very easily in Blazor Server Side :
1. Add Claim: await _userManager.AddClaimAsync(user, new Claim("LastName", user.LastName));
2. Get Claim from context: (@context.User.Claims.SingleOrDefault(x => x.Type == "FirstName").Value)
Lot of simplicity!
As long as I see there is needed additional steps in WebAssembly to get the same result, so I upvote that this feature be implemented in client side.

**Comment by guardrex** [comment]
Hello @jvallejoa ... That's a bit off topic here. `UserManager` isn't supported (and probably won't ever be) in WASM apps. Besides, we only work on docs here. Feature requests have to go to management and engineers on the product unit's repo.

**Comment by ADefWebserver** [comment]
I used the sampe code from this thread to create a Blog post: [Implementing Roles In Blazor WebAssembly](http://blazorhelpwebsite.com/Blog/tabid/61/EntryId/4376/Implementing-Roles-In-Blazor-WebAssembly.aspx)

**Comment by javiercn** [comment]
I got a working sample here https://github.com/javiercn/BlazorAuthRoles

@guardrex we need to update the topic to include using a custom `accountclaimsprincipalfactory` as multiple roles will be mapped into an array

**Comment by guardrex** [comment]
@javiercn My version is broken ... https://github.com/dotnet/AspNetCore.Docs/pull/18278.

I'll look at your approach and see if it resolves my problem there. I have time today (Monday) to address it on that existing PR.

**Comment by guardrex** [comment]
TL;DR the problem is that for a single role you get a string ... it seems to break the mapping. Idk how to fix it.

**Comment by rubeesh** [comment]
> I used the sample code from this thread to create a Blog post: [Implementing Roles In Blazor WebAssembly](http://blazorhelpwebsite.com/Blog/tabid/61/EntryId/4376/Implementing-Roles-In-Blazor-WebAssembly.aspx)

It doesnt work if more than one role is assigned to the user. For single user works fine.
Claim contains all the roles. For server side the autherization works fine.

**Comment by javiercn** [comment]
> TL;DR the problem is that for a single role you get a string (not a JSON array) ... it seems to break the mapping. Idk how to fix it.

Yeah, I realized that. I'm updating my sample.

**Comment by guardrex** [comment]
Cool ..... I'll keep an :eye: out for the update, fix up the PR, and then merge it today.

**Comment by rubeesh** [comment]
> [EDIT by guardrex to add the topic metadata]
> 
> Feel free to dupe this issue with some other issue, I couldn't find a specific one.
> 
> Content is here:
> 
> ## Steps for using roles with ApiAuthorization + Client-side Blazor
> * Configure Identity to use roles by calling AddRoles
>   ```cs
>   services.AddDefaultIdentity<ApplicationUser>(options => options.SignIn.RequireConfirmedAccount = true)
>       .AddRoles<IdentityRole>()
>       .AddEntityFrameworkStores<ApplicationDbContext>();
>   ```
> * Configure identity server to put the role claim into the id token and the access token and prevent the default mapping for roles in the JwtSecurityTokenHandler.
>   ```cs
>   services.AddIdentityServer()
>       .AddApiAuthorization<ApplicationUser, ApplicationDbContext>(options => {
>           options.IdentityResources["openid"].UserClaims.Add("role");
>           options.ApiResources.Single().UserClaims.Add("role");
>       });
>   
>   // Need to do this as it maps "role" to ClaimTypes.Role and causes issues
>   JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Remove("role");
>   ```
> * On your blazor application use `[Authorize(Roles = "admin")]` or any other role your app defines.
> * On your protected resource APIs use `[Authorize(Roles = "admin")]` or any other role your app defines.
> 
> #### Document Details
> ⚠ _Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking._
> 
> * ID: 17515ff7-32a6-c6f7-0425-535775cd6e28
> * Version Independent ID: 1416abd7-0b23-ca93-62e7-59d2532d4275
> * Content: [Secure an ASP.NET Core Blazor WebAssembly hosted app with Identity Server](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-identity-server?view=aspnetcore-3.1#feedback)
> * Content Source: [aspnetcore/security/blazor/webassembly/hosted-with-identity-server.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/security/blazor/webassembly/hosted-with-identity-server.md)
> * Product: **aspnet-core**
> * Technology: **aspnetcore-security**
> * GitHub Login: @guardrex
> * Microsoft Alias: **riande**

It works well for single role.
If more than one role is assigned Authorization doesnt work.
Token claim contains all the roles. 

For server side the autherization works fine.

**Comment by guardrex** [comment]
@rubeesh It works with a factory client-side. See the PR that I linked but wait for Javier to fix things up to cover all of the use cases.

* No roles .... I set the `Roles` prop to an empty array so that it doesn't break when trying to iterate over it in a `foreach`.
* One role ... A string value ... :boom: ... but Javier will provide the solution for this case in a bit.
* Multiple roles ... A JSON array ... the factory pattern works.

We should have the doc updated live by the EOD if all goes well.

**Comment by javiercn** [comment]
@guardrex I've updated the sample, turned out to be a bit trickier due to the different representations for 1, vs multiple role claims.

**Comment by rubeesh** [comment]
> @guardrex I've updated the sample, turned out to be a bit trickier due to the different representations for 1, vs multiple role claims.

Thanks you. Worked perfectly.

**Comment by ADefWebserver** [comment]
> @guardrex I've updated the sample, turned out to be a bit trickier due to the different representations for 1, vs multiple role claims.

@javiercn I will update my [Blog post](http://blazorhelpwebsite.com/Blog/tabid/61/EntryId/4376/Implementing-Roles-In-Blazor-WebAssembly.aspx) to follow [your sample](https://github.com/javiercn/BlazorAuthRoles).

**Comment by ADefWebserver** [comment]
@javiercn I examined [your example](https://github.com/javiercn/BlazorAuthRoles/), compared to my current [Blog post](http://blazorhelpwebsite.com/Blog/tabid/61/EntryId/4376/Implementing-Roles-In-Blazor-WebAssembly.aspx) (that uses your previous example code) and it appears the change is that this is added (and also referenced in the program.cs file): https://github.com/javiercn/BlazorAuthRoles/blob/master/Client/RolesClaimsPrincipalFactory.cs ?

**Comment by javiercn** [comment]
Yep, mostly to handle the string vs array for one to many roles.

**Comment by rubeesh** [comment]
How can i add extra information like employee id, name to claims
